### PR TITLE
Add fp16 support to word_language_model and imagenet

### PR DESCRIPTION
**Pull request by csarofeen**

### CODE REVIEW

**Comment by csarofeen** [comment]
Had a mistake which broke distributed. Fixed now.

File: imagenet/fp16util.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+def set_grad(params, params_with_grad):
+
+    for param, param_w_grad in zip(params, params_with_grad):
+        if param.grad is None:
+            param.grad = torch.nn.Parameter(param.data.new().resize_(*param.data.size()))
+        param.grad.data.copy_(param_w_grad.grad.data)
+
+
+def replace_with_BN16():
```
Comment (apaszke) [inline imagenet/fp16util.py]: No, let's not add such hacks to examples

File: imagenet/main.py [hunk 2]
```diff
@@ -200,16 +232,29 @@ def train(train_loader, model, criterion, optimizer, epoch):
...
+        loss = loss*args.loss_scale
         # compute gradient and do SGD step
-        optimizer.zero_grad()
-        loss.backward()
-        optimizer.step()
+
+        if args.fp16:
+            model.zero_grad()
+            loss.backward()
+            if args.loss_scale != 1:
```
Comment (apaszke) [inline imagenet/main.py]: You could just do `(loss / args.loss_scale).backward()` and it would be much faster. BTW it's weird that it's in fact _inverse scaling_ of the loss

File: imagenet/fp16util.py (hunk 1 above)
Comment (ngimel) [inline imagenet/fp16util.py]: 1) fp16 bn in core is currently broken
2) even if it were not, for convergence you need fp32 parameters and buffers.

File: imagenet/main.py (hunk 2 above)
Comment (csarofeen) [inline imagenet/main.py]: Loss scaling is to help maintain a well behaved range during back propagation, preventing gradients from getting truncated to 0. Running `(loss / args.loss_scale).backward()` would be the same as not having loss scaling.

**Comment by jekbradbury** [comment]
The master copy/update strategy is essentially mandatory for FP16, so there should probably be an "official" way to do it that's compatible with builtin optimizers?

**Comment by ngimel** [comment]
Imagenet example shows how to use built-in optimizer on a master copy of parameters. It is hacky, but I don't know how palatable it would be to hide all this in the built-in optimizers.
